tensorpack/graph_builder/distributed.py

```diff
@@ -277,27 +295,99 @@ def build(self, get_grad_fn, get_opt_fn):
         get_opt_fn = memoized(get_opt_fn)
         # Build the optimizer first, before entering any tower.
         # This makes sure that learning_rate is a global variable (what we expect)
-        get_opt_fn()    # TODO get_opt_fn called before main graph was built
+        get_opt_fn()  # TODO get_opt_fn called before main graph was built
 
         # Ngpu * Nvar * 2
-        grad_list = DataParallelBuilder.build_on_towers(
+        self.grad_list = DataParallelBuilder.build_on_towers(
             self.towers, get_grad_fn,
             devices=self.raw_devices,
             use_vs=[True] * len(self.towers))  # open vs at each tower
-        DataParallelBuilder._check_grad_list(grad_list)
-
-        avg_grads = aggregate_grads(
-            grad_list, colocation=False, devices=self.raw_devices)
-        with tf.device(self.param_server_device):
-            ps_var_grads = DistributedReplicatedBuilder._apply_shadow_vars(avg_grads)
-            var_update_ops = self._apply_gradients_and_copy(
-                get_opt_fn(), grad_list, ps_var_grads)
-            self._shadow_vars = [v for (__, v) in ps_var_grads]
-            self._shadow_model_vars = DistributedReplicatedBuilder._shadow_model_variables(self._shadow_vars)
+        DataParallelBuilder._check_grad_list(self.grad_list)
+
+        if self.aggregation_frequency > 1:
+            # Create the variables which will be used to store aggregated updates.
+            for k in range(len(self.gpu_shadow_vars), len(self.grad_list)):
```

Comment:
Yeah it would only happen if build is called more than once. I am not sure under what scenario this would occur(and if our code would work), so going to just put an assert here checking that `self.gpu_shadow_vars` are empty before we create them.